tripl/bio.py

```diff
@@ -0,0 +1,137 @@
+    to map e.g. column headers of a table (CSV format) into tripl semantics
+    like "namespace.entity:attribute" as well as adding the corresponding tripl
+    type.
+
+    Note that any e.g. column header that is not included in the attr_map is
+    ignored, so we can select interesting columns in that way. If the attr_map
+    specifies a key that is not found in the data, None is returned.
+
+    Example:
+
+    data = {
+        'id': 4,
+        'foo': 'bar',  # will not be included in selection
+        'day1': '2017-06-07', 'day2': '2017-06-08',
+        'time1': '13:17', 'time2': '13:21',
+        'seq': 'ACTG'
+        }
+
+    attr_map = {
+        'seq:id': 'id',
+        'seq:notpresent': 'foobar',  # key not present returns None
+        'seq:string': 'seq',
+        'seq:date': [
+            {'date:day': 'day1', 'date:time': 'time1'},
+            {'date:day': 'day2', 'date:time': 'time2'}
+        ]}
+
+    _traverse_modify(data, attr_map, ns='toy')
+    '''
+    def transformer(value):
+
+        if isinstance(value, list):
+            return value
+
+        if isinstance(value, dict):
+            entity = set()
+            vc = value.copy()  # vc .. value copy
+            # copy because iterate + mutate = bad idea, stackoverflow, 3346696
+            for i in value.keys():
+                entity.add(i.split(':')[0])
+                vc[ns + '.' + i] = value[i]
+                del vc[i]
+
+            assert len(entity) == 1, \
+                'The keys in the attribute map (obj) suggest heterogenous types.'
+
+            vc.update({ns + ':type': ns + '.type:' + entity.pop()})
+            return vc
+        else:
+            return data.get(value, None)
+
+    return _traverse(obj, callback=transformer)
+
+
+def load_csv(fp, attr_map, ns):
+    '''Turn CSV cells into triples.
+    A helper function to turn data in CSV format into entity-attribute-value
+    triples. The main ingredient that the user provides is an attribute map.
+    This map serves to map e.g. column headers of a table (CSV format) into
+    tripl semantics like "namespace.entity:attribute" as well as adding the
+    corresponding tripl type. This might seem cumbersome but note that the CSV
+    header is ambiguous about the data model.
+
+    Example:
+
+    import csv
+    from tripl import tripl as t3
+
+    fp = 'data/toy.csv'
+    namespace = 'toy'
+
+    attr_map = {
+        'seq:id': 'id',
+        'seq:virus': 'virus',
+        'seq:notpresent': 'foobar',  # key not present returns None
+        'seq:geo': 'geo',
+        'seq:date': [
+            {'date:day': 'date', 'date:time': 'time', 'date:id': 'date_id'}
+        ],
+        'seq:sample': [
+            {'sample:id': 'sample'}
+        ]
+        }
+
+    # create a triple generator
+    gen = load_csv(fp, attr_map, namespace)
+
+    # entries ought to look like this:
+    # {'toy.seq:date': [{'toy.date:day': '2017-06-01',
+    #    'toy.date:id': 't1',
```

Comment:
Maybe indent here; Otherwise, this looks great! Nicely done! I like the data api of the attr_map. It mirrors the pull data api, which is nice. I like doing that and reinforcing that as much as possible.